### ISSUE TITLE
Large XML - out of memory - C14NSerialize

### ISSUE THREAD

**Issue by romapres2010** [lestrrat-go/libxml2#60]
I try to parse and canonize quite quite big XML - about 100 Mb.

Code  is simple 
func _ProcessCanonicalizationC14n(inputXML []byte, transformXML string, withComments bool) ([]byte, error) {
doc, err := libxml2.Parse(inputXML)
	if doc != nil {
		defer doc.Free()
	}
	if err != nil {
		return nil, myerror.New("4444", errM, "libxml2.Parse()", "")
	}
	outputXML, err := dom.C14NSerialize{Mode: dom.C14NExclusive1_0, WithComments: withComments}.Serialize(doc)
	if err != nil {
		return nil, myerror.New("4444", errM, "dom.C14NSerialize()", "")
	}
	return []byte(outputXML), nil
}

When I try to bench this code in parallel (8 core), after about 100 repetition a have always got an error 
FAIL
parser error : Memory allocation failed
C14N error : Memory allocation failed : coping canonicanized document
internal buffer error : Memory allocation failed : growing buffer

Benchmark code:
	b.RunParallel(func(pb *testing.PB) {
		for pb.Next() { 
			_, err := _ProcessCanonicalizationC14n(testXML, "", false)
			if err != nil {
				b.Errorf("\n _ProcessCanonicalizationC14n error: %v", fmt.Sprintf("%+v", err))
				return
			}
		}
	})

**Comment by lestrrat** [comment]
I'm not going to just say "there can't be any bugs in my code!", but my first reaction is:

* each _ProcessCanonicalizationC14n then creates ~100MB in memory via libxml2 (the underlying C library).
* each time you receive a ~100MB Go string via Serialize()
* []byte(string) allocates for the return, so you spend another ~100MB.

I guess that means that each goroutine will be using somewhere about 300MB. You are running it over 8 cores, so that's potentially 2.4GB. That is a fair bit of memory, which doesn't surprise me that it finds errors like you did. Maybe use a smaller document, and run the benchmark longer to see if the same thing happens? if it persistent regardless of the input size, I'm sure we have a leak somewhere.

Now, to check if we have a memory leak the Go libxml2 library, I'd appreciate it if you could isolate the possible leak to which one of parse/canonicalize/serialize methods.

**Comment by romapres2010** [comment]
Here is memory consumption with Serialize [https://yadi.sk/i/8dP_KzeYsXr60g](url)

I try to exclude Serialize and memory consumption became linear [https://yadi.sk/i/stZvOegDtygzBw](url)

**Comment by lestrrat** [comment]
Sounds fair! I spotted a possible solution:

```diff
diff --git a/clib/clib.go b/clib/clib.go
index daefd86..9463d18 100644
--- a/clib/clib.go
+++ b/clib/clib.go
@@ -1812,6 +1812,7 @@ func XMLC14NDocDumpMemory(d PtrSource, mode int, withComments bool) (string, err
                e := C.MY_xmlLastError()
                return "", errors.New("c14n dump failed: " + C.GoString(e.message))
        }
+       defer C.MY_xmlFree(unsafe.Pointer(result))
        return xmlCharToString(result), nil
 }
```

Can you please try this out?

**Comment by romapres2010** [comment]
Thanks it helped.
Now memory consumption became linear with Serialize.

Intresting things. 
I test 2 variant of my function 
1. string -> ParseString -> Serialize -> return string
2. []byte -> Parse -> Serialize -> return []byte(string)

Benchmark of this two variant was almost identical.
And real memory consumption in Windows 10 monitor was identical.
Benchmark_ProcessCanonicalizationC14nString-4   	      50	 610760104 ns/op	   0.33 MB/s	26288574 B/op	      11 allocs/op
Benchmark_ProcessCanonicalizationC14n-4   	      50	 615600018 ns/op	   0.32 MB/s	77628069 B/op	      14 allocs/op

**Comment by lestrrat** [comment]
Well, doing everything in C *will* be faster, so it's not surprising to find going through a Go library would be slower? 

If you think this slowdown is because of libxml2 and not beevik/etree, then please open a new issue! I will merge a97695c and close this issue for now
